<a href="https://colab.research.google.com/github/BasmaAfreen/Stats-DeepLearning_Projects/blob/main/Association__rules.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Preprocessing:
Pre-process the dataset to ensure it is suitable for Association rules, this may include handling missing values, removing duplicates, and converting the data to appropriate format.  


In [ ]:
import pandas as pd
df=pd.read_excel("Online retail.xlsx",header=None)  # header=None bcz  by default it will consider 1st row as column names, so i have mention that this csv does not have and headrs
df

0
0     shrimp,almonds,avocado,vegetables mix,green gr...
1                                burgers,meatballs,eggs
2                                               chutney
3                                        turkey,avocado
4     mineral water,milk,energy bar,whole wheat rice...
...                                                 ...
7496                      butter,light mayo,fresh bread
7497  burgers,frozen vegetables,eggs,french fries,ma...
7498                                            chicken
7499                                 escalope,green tea
7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

[7501 rows x 1 columns]

In [ ]:
# Split the column with comma-separated values into multiple columns
df_split = df[0].str.split(',', expand=True)    #splits the values in the first column by commas and expands them into separate columns.

#df[0] :-This accesses the first column of the DataFrame df. Here, df[0] is assumed to contain strings with comma-separated values.
#.str  :- string methods allows us to apply string operations to each element in the Series.
#.split(','): _This splits each string in the Series based on the specified delimiter, which is a comma (,). It converts each string into a list of substrings.
# expand=True:- Without expand=True, the result would be a Series where each element is a list of substrings.

# Display the transformed DataFrame
df_split

0                  1            2                 3   \
0            shrimp            almonds      avocado    vegetables mix   
1           burgers          meatballs         eggs              None   
2           chutney               None         None              None   
3            turkey            avocado         None              None   
4     mineral water               milk   energy bar  whole wheat rice   
...             ...                ...          ...               ...   
7496         butter         light mayo  fresh bread              None   
7497        burgers  frozen vegetables         eggs      french fries   
7498        chicken               None         None              None   
7499       escalope          green tea         None              None   
7500           eggs    frozen smoothie  yogurt cake    low fat yogurt   

                4                 5     6               7             8   \
0     green grapes  whole weat flour  yams  cottage cheese  energy drink   
1             None              None  None            None          None   
2             None              None  None            None          None   
3             None              None  None            None          None   
4        green tea              None  None            None          None   
...            ...               ...   ...             ...           ...   
7496          None              None  None            None          None   
7497     magazines         green tea  None            None          None   
7498          None              None  None            None          None   
7499          None              None  None            None          None   
7500          None              None  None            None          None   

                9               10         11     12     13             14  \
0     tomato juice  low fat yogurt  green tea  honey  salad  mineral water   
1             None            None       None   None   None           None   
2             None            None       None   None   None           None   
3             None            None       None   None   None           None   
4             None            None       None   None   None           None   
...            ...             ...        ...    ...    ...            ...   
7496          None            None       None   None   None           None   
7497          None            None       None   None   None           None   
7498          None            None       None   None   None           None   
7499          None            None       None   None   None           None   
7500          None            None       None   None   None           None   

          15                 16               17       18         19  
0     salmon  antioxydant juice  frozen smoothie  spinach  olive oil  
1       None               None             None     None       None  
2       None               None             None     None       None  
3       None               None             None     None       None  
4       None               None             None     None       None  
...      ...                ...              ...      ...        ...  
7496    None               None             None     None       None  
7497    None               None             None     None       None  
7498    None               None             None     None       None  
7499    None               None             None     None       None  
7500    None               None             None     None       None  

[7501 rows x 20 columns]

In [ ]:
#The command pip install apyori is used to install the apyori package, which is a simple implementation of the Apriori algorithm for association rule learning.
#Apriori can be used to analyze customers' purchase patterns and discover which products are frequently bought together.
#pip install apriori

In [ ]:
trans= []
for i in range(0, 7501):
  trans.append([str(df_split.values[i,j]) for j in range(0, 20)])  # trans list will have  total 7501 sublists
trans[0]


['shrimp',
 'almonds',
 'avocado',
 'vegetables mix',
 'green grapes',
 'whole weat flour',
 'yams',
 'cottage cheese',
 'energy drink',
 'tomato juice',
 'low fat yogurt',
 'green tea',
 'honey',
 'salad',
 'mineral water',
 'salmon',
 'antioxydant juice',
 'frozen smoothie',
 'spinach',
 'olive oil']

# Association Rule Mining:
-	Implement an Apriori algorithm using tool like python with libraries such as Pandas and Mlxtend etc.
-	 Apply association rule mining techniques to the pre-processed dataset to discover interesting relationships between products purchased together.
-	 Set appropriate threshold for support, confidence and lift to extract meaning full rules.


In [ ]:
from apyori import apriori
rules = apriori(transactions = trans,  # passing  trans(transformed list)  to transaction parameter in apriori package
                min_support = 0.003,
                min_confidence = 0.2, #30%
                min_lift = 3,
                min_length = 2,
                max_length = 2)

rules # the combinations of iteams

<generator object apriori at 0x00000289485E27A0>

In [ ]:
#to get results   use   list(rules)
results=list(rules)
results

[RelationRecord(items=frozenset({'light cream', 'chicken'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]),
 RelationRecord(items=frozenset({'mushroom cream sauce', 'escalope'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]),
 RelationRecord(items=frozenset({'escalope', 'pasta'}), support=0.005865884548726837, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]),
 RelationRecord(items=frozenset({'honey', 'fromage blanc'}), support=0.003332888948140248, ordered_statistics=[OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0

In [ ]:
len(results)



9

In [ ]:
results[0]


RelationRecord(items=frozenset({'light cream', 'chicken'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)])

***This association rule suggests that "light cream" and "chicken" often appear together in transactions. With a support of 0.45%, the confidence is
roughly 29.06%, meaning that if "light cream" is purchased, there's a 29.06% chance "chicken" will be too, and the lift of 4.84 indicates this
relationship is 4.84 times more likely than by random chance.***



In [ ]:
#  accessing each values in sublists
results[0][1] # support
results[0][2][0][0] # base item
results[0][2][0][1] # add item
results[0][2][0][2] # confidence
results[0][2][0][3] # lift

4.84395061728395

##### creating a new  list for each element of results list

In [ ]:
A = []
B = []
S = []
C = []
L = []

for i in range(0,9):
    A.append(results[i][2][0][0]) # base item
    B.append(results[i][2][0][1]) # add item
    S.append(results[i][1]) # support
    C.append(results[i][2][0][2]) # confidence
    L.append(results[i][2][0][3]) # lift


#### converting list into dataframe

In [ ]:
A = pd.DataFrame(A)
B = pd.DataFrame(B)
S = pd.DataFrame(S)
C = pd.DataFrame(C)
L = pd.DataFrame(L)

data = pd.concat([A,B,S,C,L],axis=1)
data.columns = ['Baseitem','Additem','Support','confidence','Lift']
data

Baseitem      Additem   Support  confidence      Lift
0           light cream      chicken  0.004533    0.290598  4.843951
1  mushroom cream sauce     escalope  0.005733    0.300699  3.790833
2                 pasta     escalope  0.005866    0.372881  4.700812
3         fromage blanc        honey  0.003333    0.245098  5.164271
4         herb & pepper  ground beef  0.015998    0.323450  3.291994
5          tomato sauce  ground beef  0.005333    0.377358  3.840659
6           light cream    olive oil  0.003200    0.205128  3.114710
7     whole wheat pasta    olive oil  0.007999    0.271493  4.122410
8                 pasta       shrimp  0.005066    0.322034  4.506672

# Analysis and Interpretation
- •   	Analyse the generated rules to identify interesting patterns and relationships between the produc-
-   
•	Interpret the results and provide insights into customer purchasing behaviour based on the discovered rules.


Here’s a brief interpretation of each association rule in your table:

1. **light cream → chicken**:
   - Support: 0.004533 (0.45% of transactions include both items)
   - Confidence: 29.06% (When light cream is bought, there's a 29.06% chance chicken is bought too)
   - Lift: 4.84 (The likelihood of buying chicken with light cream is 4.84 times higher than chance)

2. **mushroom cream sauce → escalope**:
   - Support: 0.005733 (0.57% of transactions include both items)
   - Confidence: 30.07% (When mushroom cream sauce is bought, there's a 30.07% chance escalope is bought too)
   - Lift: 3.79 (The likelihood of buying escalope with mushroom cream sauce is 3.79 times higher than chance)

3. **pasta → escalope**:
   - Support: 0.005866 (0.59% of transactions include both items)
   - Confidence: 37.29% (When pasta is bought, there's a 37.29% chance escalope is bought too)
   - Lift: 4.70 (The likelihood of buying escalope with pasta is 4.70 times higher than chance)

4. **fromage blanc → honey**:
   - Support: 0.003333 (0.33% of transactions include both items)
   - Confidence: 24.51% (When fromage blanc is bought, there's a 24.51% chance honey is bought too)
   - Lift: 5.16 (The likelihood of buying honey with fromage blanc is 5.16 times higher than chance)

5. **herb & pepper → ground beef**:
   - Support: 0.015998 (1.60% of transactions include both items)
   - Confidence: 32.35% (When herb & pepper is bought, there's a 32.35% chance ground beef is bought too)
   - Lift: 3.29 (The likelihood of buying ground beef with herb & pepper is 3.29 times higher than chance)

6. **tomato sauce → ground beef**:
   - Support: 0.005333 (0.53% of transactions include both items)
   - Confidence: 37.74% (When tomato sauce is bought, there's a 37.74% chance ground beef is bought too)
   - Lift: 3.84 (The likelihood of buying ground beef with tomato sauce is 3.84 times higher than chance)

7. **light cream → olive oil**:
   - Support: 0.003200 (0.32% of transactions include both items)
   - Confidence: 20.51% (When light cream is bought, there's a 20.51% chance olive oil is bought too)
   - Lift: 3.11 (The likelihood of buying olive oil with light cream is 3.11 times higher than chance)

8. **whole wheat pasta → olive oil**:
   - Support: 0.007999 (0.80% of transactions include both items)
   - Confidence: 27.15% (When whole wheat pasta is bought, there's a 27.15% chance olive oil is bought too)
   - Lift: 4.12 (The likelihood of buying olive oil with whole wheat pasta is 4.12 times higher than chance)

9. **pasta → shrimp**:
   - Support: 0.005066 (0.51% of transactions include both items)
   - Confidence: 32.20% (When pasta is bought, there's a 32.20% chance shrimp is bought too)
   - Lift: 4.51 (The likelihood of buying shrimp with pasta is 4.51 times higher than chance)

These interpretations highlight the strength and likelihood of associations between the base items and the add-on items in your dataset

In [ ]:
- **this combination have highest support means that this pair appears together in approximately 1.60% of all transactions,) herb & pepper	ground beef	0.015998**
-  **the customers who came to purchase tomato sauce, 37 % chance is there that they will also buy  ground beef.**
-  **fromage blanc	is lifting honey  5.164271 times more**